### PR TITLE
refactor: parallelize `object_integration_test`

### PR DESCRIPTION
**Pull request by mr-salty**
Split the longer-running tests of `object_integration_test.cc` into
their own files, so we can run them in parallel.

The longest component test executes for only half the time of what
Carlos reported (115s vs 230s), although my timings were generally
slower than his to begin with. Also note #3872 has a drastic effect
on the two longest running tests, if we address that the runtime
should drop to ~47 seconds which is a 5x speedup.

Part of #3854

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/3873)

### CODE REVIEW

**Comment by mr-salty** [comment]
*Note there are no actual code changes here, it's a straight refactoring.*

The common bits got pulled to `testing/` and the rest are just tests from `storage_object_test.cc` split out.

I tried to fix up the #includes but we could really use IWYU (I think there were both missing and extra headers in the original).

File: google/cloud/storage/testing/object_integration_test.cc [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  if (!res) {
+    return res;
+  }
+  if (*res < 3) {
+    return Status(StatusCode::kInternal,
+                  "Expected at least three entries in /proc/self/fd: ., .., "
+                  "and the directory itself, found " +
+                      std::to_string(*res));
+  }
+  return *res - 3;
```
Comment (devbww) [inline google/cloud/storage/testing/object_integration_test.cc:65]: I'm not sure why we bother with the `3` business given that we only deal in differentials, but it should probably just be `2` and forget about "the directory itself" bit.  It is, after all, an open file.

File: google/cloud/storage/testing/object_integration_test.cc [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  dirent** namelist;
+  int res = scandir(path.c_str(), &namelist, nullptr, alphasort);
+  if (res < 0) {
+    return Status(StatusCode::kInternal, "Failed to open directory \"" + path +
+                                             "\": " + strerror(errno));
+  }
+  for (int i = 0; i < res; ++i) {
+    free(namelist[i]);
+  }
+  free(namelist);
```
Comment (devbww) [inline google/cloud/storage/testing/object_integration_test.cc:43]: It looks like a `readdir()` loop would be simpler here, and avoid extra allocations and unnecessary sorting.

**Comment by mr-salty** [comment]
since this is all existing code, my preference would be to check this in and address the comments in follow-up PRs - SG?

**Comment by mr-salty** [comment]
___
*[google/cloud/storage/tests/object_compose_many_integration_test.cc, line 48 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/3873#-M5SALLZBplDhjWeoa5t:-M5SaD9y42CIdMrg9P-K:bxq3w51) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/344ce74d941ff2f91a356d711c90ed97cf3f3584/google/cloud/storage/tests/object_compose_many_integration_test.cc#L48)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

FYI: I think this could be smaller. I think something like 33 (or 65) is Okay. Basically we want something that:

- Is larger than 32 because <= 32 results in a single `Compose()` call.
- It is not a multiple of 32 so we verify that we can handle remainders, when the size does not fit in exactly N `Compose()` calls.
- We may want to test with more than one compose, suggesting >= 64
- We want to test the case where of 32*N + 1

Hmmm, maybe we want to test with 33 and 34 to cover the cases where the remainder both requires and does not require a compose call.  OTOH, the unit tests cover some of this I believe.


</blockquote></details>

Addressed in #3875 (although I'm not clear if you actually landed on 33 so LMK)

**Comment by mr-salty** [comment]
___
*[google/cloud/storage/testing/object_integration_test.cc, line 43 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/3873#-M5PqvZ0bHdI2S98Zfxg-r1-43:-M5T95rF2TclemnFHmuA:bept859) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/344ce74d941ff2f91a356d711c90ed97cf3f3584/google/cloud/storage/testing/object_integration_test.cc#L43)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

It looks like a `readdir()` loop would be simpler here, and avoid extra allocations and unnecessary sorting.
</blockquote></details>

Fixed in #3876

**Comment by mr-salty** [comment]
___
*[google/cloud/storage/testing/object_integration_test.cc, line 65 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/3873#-M5PqvZ0bHdI2S98Zfxg-r1-65:-M5T97pC6k1xG8k4FL2W:baiz7qt) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/344ce74d941ff2f91a356d711c90ed97cf3f3584/google/cloud/storage/testing/object_integration_test.cc#L65)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

I'm not sure why we bother with the `3` business given that we only deal in differentials, but it should probably just be `2` and forget about "the directory itself" bit.  It is, after all, an open file.
</blockquote></details>

I think 3 is correct, see my comment in #3876

**Comment by mr-salty** [comment]
___
*[google/cloud/storage/tests/object_plenty_clients_serially_integration_test.cc, line 60 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/3873#-M5PqvZ0bHdI2S98Zfxm-r1-60:-M5T9H8_01NcfHmUHJik:bpl7pw7) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/344ce74d941ff2f91a356d711c90ed97cf3f3584/google/cloud/storage/tests/object_plenty_clients_serially_integration_test.cc#L60)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Should this be inside the conditional?
</blockquote></details>

yes, I also changed this around a bit in #3876 - LMK what you think

**Pull request merged**